src/channeldef.ts

```diff
@@ -75,40 +76,53 @@ export type ValueDefWithCondition<F extends FieldDef<any>, V extends ValueOrGrad
   /**
    * A field definition or one or more value definition(s) with a selection predicate.
    */
-  condition?: Conditional<F> | Conditional<ValueDef<V>> | Conditional<ValueDef<V>>[];
+  condition?: Conditional<F> | ValueOrSignalCondition<V>;
 };
 
-export type StringValueDefWithCondition<F extends Field, T extends Type = StandardType> = ValueDefWithCondition<
-  MarkPropFieldDef<F, T>,
-  string | null
->;
+export type SignalRefWithCondition<F extends FieldDef<any>, V extends ValueOrGradientOrText = Value> = SignalRef & {
+  /**
+   * A field definition or one or more value definition(s) with a selection predicate.
+   */
+  condition?: Conditional<F> | ValueOrSignalCondition<V>;
+};
 
-export type ColorGradientValueDefWithCondition<F extends Field, T extends Type = StandardType> = ValueDefWithCondition<
-  MarkPropFieldDef<F, T>,
-  Gradient | string | null
->;
+export type ValueOrSignalWithCondition<F extends FieldDef<any>, V extends ValueOrGradientOrText = Value> =
+  | ValueDefWithCondition<F, V>
+  | SignalRefWithCondition<F, V>;
+
+export type StringValueOrSignalWithCondition<
+  F extends Field,
+  T extends Type = StandardType
+> = ValueOrSignalWithCondition<MarkPropFieldDef<F, T>, string | null>;
 
-export type NumericValueDefWithCondition<F extends Field> = ValueDefWithCondition<
+export type ColorGradientValueOrSignalWithCondition<
+  F extends Field,
+  T extends Type = StandardType
+> = ValueOrSignalWithCondition<MarkPropFieldDef<F, T>, Gradient | string | null>;
+
+export type NumericValueOrSignalWithCondition<F extends Field> = ValueOrSignalWithCondition<
   MarkPropFieldDef<F, StandardType>,
   number
 >;
 
 export type TypeForShape = 'nominal' | 'ordinal' | 'geojson';
 
-export type ShapeValueDefWithCondition<F extends Field> = StringValueDefWithCondition<F, TypeForShape>;
+export type ShapeValueOrSignalWithCondition<F extends Field> = StringValueOrSignalWithCondition<F, TypeForShape>;
```

Comment:
Question: Should we be accurate and use the name `Shape(/Color/...)ValueOrSignalWithCondition` ?

Or should we preserve the old name (e.g., `ShapeValueDefWithCondition`) as we currently hide signals?

Reply:
Can we name the type after what it is used for?

Reply:
Not sure what you mean, but thinking more,  the right thing to do is to keep the accurate name in the TS code.  

If we want to hide "signals", we should do that in the renaming process, so it's not confusing when we read TS code.